### PR TITLE
Fix UnauthorizedError on GitHub Enterprise publish

### PR DESCRIPTION
**Pull request by necosta**
When `gitHostingURL` is set, we see the following error:

```
Error making request to GitHub
 | => bat microsites.github.GitHubOps.$anonfun$run$1(GitHubOps.scala:208)
	at uncancelable @ org.http4s.client.ConnectionManager$.pool(ConnectionManager.scala:83)
	at uncancelable @ org.http4s.client.ConnectionManager$.pool(ConnectionManager.scala:83)
	at flatMap @ org.http4s.client.PoolManager.releaseRecyclable(PoolManager.scala:225)
	at uncancelable @ org.http4s.client.ConnectionManager$.pool(ConnectionManager.scala:83)
	at unsafeRunSync @ microsites.MicrositeAutoImportSettings.$anonfun$micrositeTasksSettings$14(MicrositeKeys.scala:496)
	at withPermit @ org.http4s.client.PoolManager.shutdown(PoolManager.scala:370)
	at uncancelable @ org.http4s.client.ConnectionManager$.pool(ConnectionManager.scala:83)
	at withPermit @ org.http4s.client.PoolManager.shutdown(PoolManager.scala:370)
	at guarantee$extension @ org.http4s.client.blaze.Http1Connection.$anonfun$executeRequest$2(Http1Connection.scala:193)
	at map @ fs2.internal.CompileScope.$anonfun$close$9(CompileScope.scala:246)
Caused by: UnauthorizedError(Bad credentials, https://docs.github.com/rest)

```

We need to adjust GithubConfig base/authorize/accessToken URLs to contain the `gitHostingURL` host.

---

@TomyMeren , can you:
1. Clone my repo (branch fix/github_enterprise)
2. Run publishLocal (haven't checked but should be configured)
3. On your repo, change resolver to include resolvers += Resolver.mavenLocal, adjust github4s version to this published version
4. Compile, run and inform if it's now working

---

### CODE REVIEW

File: src/main/scala/microsites/MicrositeKeys.scala [hunk 1]
```diff
@@ -551,6 +556,24 @@ trait MicrositeAutoImportSettings extends MicrositeKeys {
...
 
+  protected[this] def buildGithubConfig(hostingUrl: String)(implicit log: Logger): GithubConfig =
+    Either.catchOnly[MalformedURLException](new URL(hostingUrl)) match {
+      case Right(url) if url.getProtocol.startsWith("https") =>
+        val replaceHost: String => String = s => s.replace("github.com", url.getHost)
+        GithubConfig.default
+          .copy(
+            baseUrl = replaceHost(s"${GithubConfig.default.baseUrl}"),
+            authorizeUrl = replaceHost(s"${GithubConfig.default.authorizeUrl}"),
+            accessTokenUrl = replaceHost(s"${GithubConfig.default.accessTokenUrl}")
```
Comment (fedefernandez) [inline src/main/scala/microsites/MicrositeKeys.scala:567]:
```suggestion
            baseUrl = replaceHost(GithubConfig.default.baseUrl),
            authorizeUrl = replaceHost(GithubConfig.default.authorizeUrl),
            accessTokenUrl = replaceHost(GithubConfig.default.accessTokenUrl)
```

**Review by fedefernandez — APPROVED** [review]
Thanks @necosta. Please, @TomyMeren, let me know if this fixes the issue.

File: src/main/scala/microsites/MicrositeKeys.scala (hunk 1 above)
Comment (necosta) [reply src/main/scala/microsites/MicrositeKeys.scala]: Thanks! Fixed.

**Comment by necosta** [comment]
After reading https://github.com/google/go-github/issues/958, adjusted baseUrl logic.

**Review by fedefernandez — APPROVED** [review]

**Comment by necosta** [comment]
Added edge-case scenario for when micrositeGitHostingUrl is still pointing to public GitHub:
 
`case Right(url) if url.getHost == "github.com" => GithubConfig.default`

**Review by fedefernandez — DISMISSED** [review]

**Comment by TomyMeren** [comment]
Good progress! @necosta.

The current error is this: `NotFoundError(Not Found, https://docs.github.com/enterprise/3.4/rest)`
After Debugging I've seem that it is throwing here https://github.com/47degrees/sbt-microsites/blob/26a426e32df5fd6800628a0303d89a7118954b06/src/main/scala/microsites/github/GitHubOps.scala#L200

It could be a problem with the method `updateReference` in the github4s repo but I could't find the https://docs.github.com/enterprise/3.4/rest value in the project

**Review by TomyMeren — APPROVED** [review]
LGTM!

**Comment by necosta** [comment]
Thanks @TomyMeren for testing. I did mess up the PR when I last updated, will get the changes from the "patched" branch, adjust the tests, then I think we are good to merge. Will do it now.

**Comment by necosta** [comment]
> When I publish the sbt-microsite project locally, for use in the project, version **1.4.1**+1.....-SNAPSHOT is created. The latest release version is **1.4.2**. Do you know why this is?

Yes, publishLocal is configured to build a "SNAPSHOT" version and not a "release" version. It should work this way to avoid any version conflict.

**Comment by necosta** [comment]
Green CI. It's ready for a final review @fedefernandez , feel free to merge after you approve the PR. Thank you @TomyMeren for testing it.

File: src/main/scala/microsites/MicrositeKeys.scala [hunk 2]
```diff
@@ -551,6 +557,25 @@ trait MicrositeAutoImportSettings extends MicrositeKeys {
...
     }
 
+  protected[this] def buildGithubConfig(hostingUrl: String)(implicit log: Logger): GithubConfig =
+    Either.catchOnly[MalformedURLException](new URL(hostingUrl)) match {
+      case Right(url) if url.getHost == "github.com" => GithubConfig.default
+      case Right(url) if url.getProtocol.startsWith("https") =>
+        val replaceHost: String => String = s => s.replace("github.com", url.getHost)
+        GithubConfig.default
+          .copy(
+            baseUrl = s"${url.getProtocol}://${url.getHost}/api/v3/",
```
Comment (fedefernandez) [inline src/main/scala/microsites/MicrositeKeys.scala:567]: This is my only concern, is this always `v3`?

**Review by fedefernandez — APPROVED** [review]

File: src/main/scala/microsites/MicrositeKeys.scala (hunk 2 above)
Comment (necosta) [reply src/main/scala/microsites/MicrositeKeys.scala:567]: The research I did, REST API is only on v3 (and there is a new GraphQL API on v4)
https://github.com/orgs/community/discussions/24559
I didn't see any GitHub Enterprise mention of v2 or v4 so I didn't go ahead with soft-coding this value.
But if you think we should, I can make that change.

Comment (TomyMeren) [reply src/main/scala/microsites/MicrositeKeys.scala:567]: To the GitHub Enterprise Server versions, for Cloud version it changes to` https://api.github.com`
https://docs.github.com/en/enterprise-server@3.4/rest/quickstart
https://docs.github.com/en/enterprise-cloud@latest/rest/quickstart?apiVersion=2022-11-28

**Comment by TomyMeren** [comment]
> > When I publish the sbt-microsite project locally, for use in the project, version **1.4.1**+1.....-SNAPSHOT is created. The latest release version is **1.4.2**. Do you know why this is?
> 
> Yes, publishLocal is configured to build a "SNAPSHOT" version and not a "release" version. It should work this way to avoid any version conflict.

Sorry for my unclear comment.
I was talking about the difference between 1.4.1 and 1.4.2

**Review by fedefernandez — APPROVED** [review]

File: src/main/scala/microsites/MicrositeKeys.scala (hunk 2 above)
Comment (fedefernandez) [reply src/main/scala/microsites/MicrositeKeys.scala:567]: Could you please add an issue with the information? Thanks

**Comment by necosta** [comment]
> > > When I publish the sbt-microsite project locally, for use in the project, version **1.4.1**+1.....-SNAPSHOT is created. The latest release version is **1.4.2**. Do you know why this is?
> > 
> > 
> > Yes, publishLocal is configured to build a "SNAPSHOT" version and not a "release" version. It should work this way to avoid any version conflict.
> 
> Sorry for my unclear comment. I was talking about the difference between 1.4.1 and 1.4.2

Ah I see. I tried now, I do get 1.4.2-SNAPSHOT when publishing locally. Perhaps you are missing the git tag?

File: src/main/scala/microsites/MicrositeKeys.scala (hunk 2 above)
Comment (necosta) [reply src/main/scala/microsites/MicrositeKeys.scala:567]: I hope I'm not adding confusion, but there is no issue. When it's a GH Enterprise target it's `/api/v3` , when it's GitHub, it's `api.github.com`.
This PR handles both cases, defaulting to https://github.com/47degrees/github4s/blob/main/github4s/shared/src/main/scala/github4s/GithubConfig.scala#L43 when hostname is `github.com`.

**Review by TomyMeren — APPROVED** [review]
Yes It was a missing tag
LGTM!

**Comment by necosta** [comment]
Green CI and all comments resolved. Is it ok to merge to master @fedefernandez ? thanks

**Review by fedefernandez — APPROVED** [review]

**Pull request merged**